File: src/types/position.ts
/**
 * Position tracking types for incremental reading
 */

/**
 * Unified position representation for all document types
 */
export type DocumentPosition =
  | { type: 'page'; page: number; offset?: number }
  | { type: 'scroll'; percent: number; element_id?: string }
  | { type: 'cfi'; cfi: string; offset?: number }
  | { type: 'time'; seconds: number; total_duration?: number };

/**
 * Calculate progress percentage from a DocumentPosition
 */
export function getPositionProgress(position: DocumentPosition): number | null {
  switch (position.type) {
    case 'page':
      // Without total pages, we can't calculate percentage
      return null;
    case 'scroll':
      return position.percent;
    case 'cfi':
      // CFI doesn't directly give us progress
      return null;
    case 'time':
      if (position.total_duration) {
        return (position.seconds / position.total_duration) * 100;
      }
      return null;
  }
}

/**
 * Create a page position
 */
export function pagePosition(page: number, offset?: number): DocumentPosition {
  return { type: 'page', page, offset };
}

/**
 * Create a scroll position
 */
export function scrollPosition(percent: number, elementId?: string): DocumentPosition {
  return { type: 'scroll', percent, element_id: elementId };
}

/**
 * Create a CFI position (EPUB)
 */
export function cfiPosition(cfi: string, offset?: number): DocumentPosition {
  return { type: 'cfi', cfi, offset };
}

/**
 * Create a time position (video/audio)
 */
export function timePosition(seconds: number, totalDuration?: number): DocumentPosition {
  return { type: 'time', seconds, total_duration: totalDuration };
}

/**
 * Format position for display
 */
export function formatPosition(position: DocumentPosition): string {
  switch (position.type) {
    case 'page':
      return position.offset !== undefined
        ? `Page ${position.page} (${Math.round(position.offset * 100)}%)`
        : `Page ${position.page}`;
    case 'scroll':
      return `${position.percent.toFixed(1)}%`;
    case 'cfi':
      return `CFI: ${position.cfi.substring(0, 20)}...`;
    case 'time':
      if (position.total_duration) {
        const formatTime = (s: number) => {
          const mins = Math.floor(s / 60);
          const secs = s % 60;
          return `${mins}:${secs.toString().padStart(2, '0')}`;
        };
        return `${formatTime(position.seconds)} / ${formatTime(position.total_duration)}`;
      }
      const mins = Math.floor(position.seconds / 60);
      const secs = position.seconds % 60;
      return `${mins}:${secs.toString().padStart(2, '0')}`;
  }
}

/**
 * Bookmark for a specific position in a document
 */
export interface Bookmark {
  id: string;
  document_id: string;
  name: string;
  position: DocumentPosition;
  thumbnail?: string; // Base64 encoded thumbnail
  created_at: string;
}

/**
 * Reading session tracking
 */
export interface ReadingSession {
  id: string;
  document_id: string;
  started_at: string;
  ended_at: string | null;
  duration_seconds: number;
  pages_read: number | null;
  progress_start: number;
  progress_end: number;
}

/**
 * Daily reading statistics
 */
export interface DailyReadingStats {
  date: string; // YYYY-MM-DD
  total_seconds: number;
  documents_read: number;
  total_pages_read: number;
  session_count: number;
}

/**
 * Reading streak information
 */
export interface ReadingStreak {
  current_streak: number;
  longest_streak: number;
  last_activity_date: string | null;
  is_active: boolean;
}

/**
 * Document with progress information
 */
export interface DocumentWithProgress {
  id: string;
  progress: number;
  title: string;
  date_modified: number;
}

/**
 * Progress grouping for Continue Reading page
 */
export type ProgressGroup = 'not-started' | 'just-started' | 'halfway' | 'almost-done' | 'completed';

export function getProgressGroup(progress: number): ProgressGroup {
  if (progress === 0) return 'not-started';
  if (progress < 25) return 'just-started';
  if (progress < 75) return 'halfway';
  if (progress < 99) return 'almost-done';
  return 'completed';
}

export interface ProgressGroupInfo {
  group: ProgressGroup;
  label: string;
  icon: string;
  description: string;
}

export const PROGRESS_GROUPS: Record<ProgressGroup, ProgressGroupInfo> = {
  'not-started': {
    group: 'not-started',
    label: 'Not Started',
    icon: '📚',
    description: "Documents you haven't started reading yet",
  },
  'just-started': {
    group: 'just-started',
    label: 'Just Started',
    icon: '🔖',
    description: "Documents you've started (0-25% progress)",
  },
  'halfway': {
    group: 'halfway',
    label: 'Halfway Through',
    icon: '📖',
    description: "Documents you're making progress on (25-75%)",
  },
  'almost-done': {
    group: 'almost-done',
    label: 'Almost Done',
    icon: '🏁',
    description: "Documents you're nearly finished with (75-99%)",
  },
  'completed': {
    group: 'completed',
    label: 'Completed',
    icon: '✅',
    description: "Documents you've finished reading",
  },
};

File: src-tauri/src/models/position.rs
//! Position tracking models for incremental reading

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::fmt;

/// Unified position representation for all document types
/// This abstracts the different ways positions are tracked:
/// - PDF: page numbers
/// - EPUB: Canonical Fragment Identifiers (CFI)
/// - HTML/Markdown: scroll percentage
/// - Video: time in seconds
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum DocumentPosition {
    /// Page-based position (PDF, DJVU)
    Page {
        page: u32,
        /// Optional offset within the page (0.0 to 1.0)
        #[serde(skip_serializing_if = "Option::is_none")]
        offset: Option<f32>,
    },
    /// Scroll-based position (HTML, Markdown, plain text)
    Scroll {
        /// Percentage from 0.0 to 100.0
        percent: f32,
        /// Optional element ID for anchor-based positioning
        #[serde(skip_serializing_if = "Option::is_none")]
        element_id: Option<String>,
    },
    /// EPUB Canonical Fragment Identifier
    Cfi {
        /// The CFI string (e.g., "/body/DocFragment[2]/chap1[4]!")
        cfi: String,
        /// Optional offset within the CFI location (0.0 to 1.0)
        #[serde(skip_serializing_if = "Option::is_none")]
        offset: Option<f32>,
    },
    /// Time-based position (video, audio)
    Time {
        /// Position in seconds
        seconds: u32,
        /// Total duration in seconds (for progress calculation)
        #[serde(skip_serializing_if = "Option::is_none")]
        total_duration: Option<u32>,
    },
}

impl DocumentPosition {
    /// Calculate progress as a percentage (0.0 to 100.0)
    /// Returns None if progress cannot be determined
    pub fn progress_percent(&self) -> Option<f32> {
        match self {
            DocumentPosition::Page { page, .. } => {
                // Without total pages, we can't calculate percentage
                None
            }
            DocumentPosition::Scroll { percent, .. } => Some(*percent),
            DocumentPosition::Cfi { .. } => {
                // CFI doesn't directly give us progress
                None
            }
            DocumentPosition::Time {
                seconds,
                total_duration,
            } => {
                if let Some(total) = total_duration {
                    if *total > 0 {
                        Some((*seconds as f32 / *total as f32) * 100.0)
                    } else {
                        None
                    }
                } else {
                    None
                }
            }
        }
    }

    /// Set total for progress calculation (pages for PDF, duration for video)
    pub fn with_total(self, total: u32) -> DocumentPosition {
        match self {
            DocumentPosition::Page { page, offset } => {
                // Store total in a way we can use for progress
                DocumentPosition::Page { page, offset }
            }
            DocumentPosition::Time {
                seconds,
                total_duration: _,
            } => DocumentPosition::Time {
                seconds,
                total_duration: Some(total),
            },
            _ => self,
        }
    }

    /// Create a page position
    pub fn page(page: u32) -> Self {
        DocumentPosition::Page {
            page,
            offset: None,
        }
    }

    /// Create a page position with offset
    pub fn page_with_offset(page: u32, offset: f32) -> Self {
        DocumentPosition::Page {
            page,
            offset: Some(offset),
        }
    }

    /// Create a scroll position
    pub fn scroll(percent: f32) -> Self {
        DocumentPosition::Scroll {
            percent: percent.clamp(0.0, 100.0),
            element_id: None,
        }
    }

    /// Create a scroll position with element anchor
    pub fn scroll_with_element(percent: f32, element_id: String) -> Self {
        DocumentPosition::Scroll {
            percent: percent.clamp(0.0, 100.0),
            element_id: Some(element_id),
        }
    }

    /// Create a CFI position
    pub fn cfi(cfi: String) -> Self {
        DocumentPosition::Cfi {
            cfi,
            offset: None,
        }
    }

    /// Create a CFI position with offset
    pub fn cfi_with_offset(cfi: String, offset: f32) -> Self {
        DocumentPosition::Cfi {
            cfi,
            offset: Some(offset),
        }
    }

    /// Create a time position
    pub fn time(seconds: u32) -> Self {
        DocumentPosition::Time {
            seconds,
            total_duration: None,
        }
    }

    /// Create a time position with duration
    pub fn time_with_duration(seconds: u32, duration: u32) -> Self {
        DocumentPosition::Time {
            seconds,
            total_duration: Some(duration),
        }
    }

    /// Get the position type as a string
    pub fn type_name(&self) -> &'static str {
        match self {
            DocumentPosition::Page { .. } => "page",
            DocumentPosition::Scroll { .. } => "scroll",
            DocumentPosition::Cfi { .. } => "cfi",
            DocumentPosition::Time { .. } => "time",
        }
    }
}

impl fmt::Display for DocumentPosition {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            DocumentPosition::Page { page, offset } => {
                if let Some(off) = offset {
                    write!(f, "Page {} ({:.0}%)", page, off * 100.0)
                } else {
                    write!(f, "Page {}", page)
                }
            }
            DocumentPosition::Scroll { percent, .. } => {
                write!(f, "{:.1}%", percent)
            }
            DocumentPosition::Cfi { cfi, .. } => {
                write!(f, "CFF: {}", cfi)
            }
            DocumentPosition::Time { seconds, total_duration } => {
                if let Some(total) = total_duration {
                    write!(f, "{}:{:02} / {}:{:02}", seconds / 60, seconds % 60, total / 60, total % 60)
                } else {
                    write!(f, "{}:{:02}", seconds / 60, seconds % 60)
                }
            }
        }
    }
}

/// Bookmark for a specific position in a document
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Bookmark {
    pub id: String,
    pub document_id: String,
    pub name: String,
    pub position: DocumentPosition,
    /// Base64-encoded thumbnail image data (optional)
    pub thumbnail: Option<String>,
    pub created_at: DateTime<Utc>,
}

impl Bookmark {
    pub fn new(document_id: String, name: String, position: DocumentPosition) -> Self {
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            document_id,
            name,
            position,
            thumbnail: None,
            created_at: Utc::now(),
        }
    }
}

/// Reading session tracking
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReadingSession {
    pub id: String,
    pub document_id: String,
    pub started_at: DateTime<Utc>,
    pub ended_at: Option<DateTime<Utc>>,
    pub duration_seconds: u32,
    pub pages_read: Option<u32>,
    pub progress_start: f32,
    pub progress_end: f32,
}

impl ReadingSession {
    pub fn new(document_id: String, progress_start: f32) -> Self {
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            document_id,
            started_at: Utc::now(),
            ended_at: None,
            duration_seconds: 0,
            pages_read: None,
            progress_start,
            progress_end: progress_start,
        }
    }

    /// End the session and calculate duration
    pub fn end(&mut self, progress_end: f32) {
        self.ended_at = Some(Utc::now());
        self.progress_end = progress_end;
        let duration = Utc::now() - self.started_at;
        self.duration_seconds = duration.num_seconds() as u32;
    }

    /// Get the actual reading progress made during this session
    pub fn progress_made(&self) -> f32 {
        self.progress_end - self.progress_start
    }
}

/// Daily reading statistics for streaks and goals
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DailyReadingStats {
    pub date: String, // YYYY-MM-DD format
    pub total_seconds: u32,
    pub documents_read: u32,
    pub total_pages_read: u32,
    pub session_count: u32,
}

impl DailyReadingStats {
    /// Get reading time in minutes
    pub fn minutes(&self) -> u32 {
        self.total_seconds / 60
    }

    /// Check if this day has any reading activity
    pub fn has_activity(&self) -> bool {
        self.total_seconds > 0
    }
}
